
===========================================================
25.0 Galaxy Release (June 2025)
===========================================================

.. include:: _header.rst

Highlights
===========================================================

**Empower Users to Build More Kinds of Collections, More Intelligently**. Galaxy’s collection system has
been significantly enhanced, making it easier to create and manage collections—including mixed paired
and unpaired datasets, nested lists, and more—directly from selected datasets. New wizard-style builders
guide users step-by-step, auto-detecting data structures and offering both instant and advanced setup
options, while improved rule-based import activities, clearer feedback, and numerous refinements.

**Add ZIP explorer to import individual files from local or remote ZIP archives**. Galaxy now features a
ZIP explorer, streamlining the process of importing individual files from both local and remote ZIP
archives.

**A New Unified View for Datasets**. Galaxy 25.0 introduces a unified view for datasets,
consolidating common dataset actions into a single, streamlined interface, all accessible via the eye
icon on datasets.

**Allow Rerunning Workflows with the Same Inputs and Parameters**. Users can now easily rerun any
workflow with the exact same inputs and parameters as a previous run - with a simple click of the new
"Rerun" button in the workflow invocation view.

**An Enhanced Workflow Run Form Interface**. Galaxy 25.0 introduces a streamlined workflow run interface
that allows users to upload inputs, create collections, and configure parameters — all within a single,
unified form.

**Introducing READMEs for Workflows in the Galaxy Interface.** Galaxy 25.0 introduces support for
associating README/help text with workflows. This provides a way to document workflows directly in the
Galaxy interface, with support for Markdown formatting, links, and images. The README is a part of the
workflow object in the database.

**Bulk Actions for Workflows**. Galaxy 25.0 introduces support for selecting multiple workflows to
perform actions in bulk, streamlining workflow management for users.

**Redesigned File Sources and Storage Location Interfaces**. Galaxy 25.0 introduces a modernized
interface for accessing Remote File Sources and Storage Locations, making navigation and selection more
intuitive and user-friendly.

**Enable Cloning Subworkflows in the Workflow Editor**. In the Workflow Editor, subworkflows can now be
cloned just like regular workflow steps, with a singular click of the "Duplicate" button on a
subworkflow node.

**JupyterLite for Lightweight In-Browser Notebooks within Galaxy**. JupyterLite — a fully browser-based
Jupyter environment — is now available as a visualization (accessible right in the Galaxy center panel)
with support for interacting directly with Galaxy datasets and APIs, without requiring any local setup,
or using the Jupyter interactive tool.

**New Interactive Tools Panel in the Activity Bar**. Interactive Tools now have a dedicated panel in the
Activity Bar, making them easier to find and launch. Each tool has its own icon, and a badge shows how
many are currently running. This improves visibility and separates ITs from standard Galaxy tools.

**Preferred Visualization Display for Dataset Types**. Galaxy now allows administrators to configure default
visualizations for specific datatypes. Instead of showing raw file content in the (new) Preview tab, datasets with a
preferred visualization will now display that visualization by default.

**User-Defined Tools (Beta)**. Galaxy 25.0 introduces support for user-defined tools — a new feature that lets selected
users define and run tools directly in the Galaxy interface, without requiring regular administrative intervention.

**Job Cancellation Notifications for Admins**. Admins can now choose to notify users when canceling
jobs. Users receive grouped notifications with job details, and optional email delivery if enabled.

Please see the `25.0 user release notes <25.0_announce_user.html>`__ for a summary of new user features.

Get Galaxy
===========================================================

The code lives at `GitHub <https://github.com/galaxyproject/galaxy>`__ and you should have `Git <https://git-scm.com/>`__ to obtain it.

To get a new Galaxy repository run:
  .. code-block:: shell

      $ git clone -b release_25.0 https://github.com/galaxyproject/galaxy.git

To update an existing Galaxy repository run:
  .. code-block:: shell

      $ git fetch origin && git checkout release_25.0 && git pull --ff-only origin release_25.0

See the `community hub <https://galaxyproject.org/develop/source-code/>`__ for additional details on source code locations.


Admin Notes
===========================================================

**Removal of support for Python 3.8 in Galaxy release 25.0**
  The previous Galaxy release (24.2) was the last to support Python 3.8.
  Galaxy 25.0 and all future releases will now require Python 3.9 or later.

**Removal of Backbone-Based Chart Visualizations**
  The legacy Backbone-based chart visualization modules have been removed. These visualizations were no
  longer maintained, and the last active one ("Nora") is currently non-functional. Future visualizations
  should be restored or developed using the new script entry point approach.

  As part of this change, no visualizations will be bundled with the Galaxy client build by default.

Configuration Changes
===========================================================


Added
-----

The following configuration options are new

config/user_preferences_extra_conf.yml.sample
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  preferences.dataverse
-  preferences.dataverse_sandbox
-  preferences.elabftw

config/themes_conf.yml.sample
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  anvil.masthead
-  anvil.masthead.link
-  anvil.masthead.logo
-  anvil.masthead.text

config/galaxy.yml.sample:galaxy
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  mapping.tool_shed.mapping.canonical_email_rules
-  mapping.tool_shed.mapping.canonical_email_rules.default.all
-  mapping.tool_shed.mapping.canonical_email_rules.default.gmail.com
-  mapping.tool_shed.mapping.canonical_email_rules.default.proton.me
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+)
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+).mapping.aliases
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+).mapping.ignore_case
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+).mapping.ignore_dots
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+).mapping.sub_addressing
-  mapping.tool_shed.mapping.canonical_email_rules.mapping.regex;(.+).mapping.sub_addressing_delim
-  mapping.tool_shed.mapping.email_ban_file


Changed
-------

The following configuration options have been changed

config/file_sources_conf.yml.sample
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-   has changed from

   ::

    {'type': 'dropbox', 'id': 'dropbox1', 'label': 'Dropbox files (configure access in user preferences)', 'doc': 'Your Dropbox files - configure an access token via the user preferences', 'access_token': "${user.preferences['dropbox|access_token']}"}
    {'type': 'webdav', 'id': 'owncloud1', 'label': 'OwnCloud', 'doc': 'External OwnCloud files (configure access in user preferences)', 'url': "${user.preferences['owncloud|url']}", 'root': "${user.preferences['owncloud|root']}", 'login': "${user.preferences['owncloud|username']}", 'password': "${user.preferences['owncloud|password']}", 'temp_path': '/your/temp/path', 'writable': False}
    {'type': 'posix', 'root': '/data/5/galaxy_import/galaxy_user_data/covid-19/data/sequences/', 'id': 'covid19-raw-sequences', 'label': 'COVID-19 FASTQ', 'doc': 'COVID-19 RAW sequences in FASTQ format'}
    {'type': 'posix', 'root': '/data/db/databases/pdb/pdb/', 'id': 'pdb-gzip', 'doc': 'Protein Data Bank (PDB)', 'label': 'PDB'}
    {'type': 'ftp', 'id': 'ebi-ftp', 'label': 'EBI FTP server', 'doc': 'European Bioinformatic Institute FTP server', 'host': 'ftp.ebi.ac.uk', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ftp', 'id': 'ncbi-ftp', 'label': 'NCBI FTP server', 'doc': 'NCBI FTP server', 'host': 'ftp.ncbi.nlm.nih.gov', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ftp', 'id': 'ensembl-ftp', 'label': 'ENSEMBL FTP server', 'doc': 'ENSEMBL FTP server', 'host': 'ftp.ensemblgenomes.org/vol1/pub/', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ssh', 'id': 'writeable-ssh-dir', 'requires_roles': 'writeable-galaxy-role', 'writable': True, 'label': 'Push your files to me', 'doc': 'This is an example of a writeable SSH dir', 'host': 'coolhost', 'user': 'user', 'passwd': 'passwd', 'timeout': 10, 'path': '/home/cooluser/', 'config_path': '', 'port': 2222}
    {'type': 's3fs', 'label': 'My MinIO storage', 'endpoint_url': 'https://minio.usegalaxy.eu', 'id': 'galaxy-minio-storage', 'doc': 'Galaxy MinIO S3 storage', 'anon': False, 'secret': 'UHAJ6asd6asdhasd', 'key': 'MCJU76agdt98GGFAROIP7'}
    {'type': 's3fs', 'label': 'Genome Ark', 'id': 'genomeark', 'doc': 'Access to Genome Ark open data on AWS.', 'bucket': 'genomeark', 'anon': True}
    {'type': 's3fs', 'label': '1000 Genomes', 'id': '1000genomes', 'doc': 'Access to the 1000 Genomes Project with human genetic variation, including SNPs, structural variants, and their haplotype context.', 'bucket': '1000genomes', 'anon': True}
    {'type': 's3fs', 'label': 'The Cancer Genome Atlas', 'id': 'tcga-2-open', 'doc': 'Access to the Cancer Genome Atlas (TCGA)', 'bucket': 'tcga-2-open', 'anon': True}
    {'type': 's3fs', 'label': 'COVID-19 Data Lake', 'id': 'covid19-lake', 'doc': 'A centralized repository of up-to-date and curated datasets on or related to the spread and characteristics of the novel corona virus (SARS-CoV-2) and its associated illness, COVID-19', 'bucket': 'covid19-lake', 'anon': True}
    {'type': 's3fs', 'label': 'Encyclopedia of DNA Elements (ENCODE)', 'id': 'encode-public', 'doc': 'The Encyclopedia of DNA Elements (ENCODE) Consortium is an international collaboration of research groups funded by the National Human Genome Research Institute (NHGRI)', 'bucket': 'encode-public', 'anon': True}
    {'type': 's3fs', 'label': 'Sentinel-3', 'id': 'meeo-s3-nrt', 'doc': 'European Commission’s Copernicus Earth Observation Programme. Sentinel-3 is a polar orbiting satellite that completes 14 orbits of the Earth a day.', 'bucket': 'meeo-s3/NRT/', 'anon': True}
    {'type': 's3fs', 'label': 'Sentinel-5P Level 2', 'id': 'meeo-s5p-nrti', 'doc': 'Observations from the Sentinel-5 Precursor satellite of the Copernicus Earth Observation Programme. It contains a polar orbiting satellite that completes 14 orbits of the Earth a day.', 'bucket': 'meeo-s5p/RPRO/', 'anon': True}
    {'type': 's3fs', 'label': 'Coupled Model Intercomparison Project 6', 'id': 'esgf-world', 'doc': 'The sixth phase of global coupled ocean-atmosphere general circulation model ensemble', 'bucket': 'esgf-world', 'anon': True}
    {'type': 's3fs', 'label': 'CMIP6 GCMs downscaled using WRF', 'id': 'wrf-cmip6-noversioning', 'doc': 'High-resolution historical and future climate simulations from 1980-2100', 'bucket': 'wrf-cmip6-noversioning', 'anon': True}
    {'type': 's3fs', 'label': 'NOAA Global Forecast System (GFS)', 'id': 'noaa-gfs-bdp-pds', 'doc': 'The Global Forecast System (GFS) is a weather forecast model produced by the National Centers for Environmental Prediction (NCEP).', 'bucket': 'noaa-gfs-bdp-pds', 'anon': True}
    {'type': 's3fs', 'label': 'NOAA Unified Forecast System Subseasonal to Seasonal Prototype 5', 'id': 'noaa-ufs-prototype5-pds', 'doc': 'The Unified Forecast System Subseasonal to Seasonal prototype 5 (UFS S2Sp5) dataset is reforecast data from the UFS atmosphere-ocean.', 'bucket': 'noaa-ufs-prototype5-pds', 'anon': True}
    {'type': 's3fs', 'label': 'Copernicus Digital Elevation Model (DEM)', 'id': 'copernicus-dem-30m', 'doc': 'The Copernicus DEM is a Digital Surface Model (DSM) which represents the surface of the Earth including buildings, infrastructure and vegetation.', 'bucket': 'copernicus-dem-30m', 'anon': True}
    {'type': 'http', 'label': 'Custom http filesource', 'id': 'httpcustom', 'url_regex': '^https?://myprotectedsite.org/', 'http_headers': {'Authorization': '#import base64\nBasic ${base64.b64encode(str.encode(user.preferences[\'myprotectedsite|username\'] + ":" + user.preferences[\'myprotectedsite|password\'])).decode()}'}, 'doc': 'Only define this if you want custom control over http downloads. You should also define a stock http source (below) or only downloads from this site will be allowed.'}
    {'type': 'http', 'label': 'Stock http filesource', 'id': 'httpstock', 'doc': 'Make sure to define this generic http file source if you have defined any other http file sources, or stock http download capability will be disabled.'}
    {'type': 'drs', 'label': 'Custom DRS filesource', 'id': 'drscustom', 'url_regex': '^drs://mydrssite.org/', 'http_headers': {'Authorization': '#import base64\nBasic ${base64.b64encode(str.encode(user.preferences[\'mydrssite|username\'] + ":" + user.preferences[\'mydrssite|password\'])).decode()}'}, 'doc': 'Define this if you want custom control over drs downloads. You should also define a stock drs source (below) or only downloads from this drs server will be allowed.'}
    {'type': 'drs', 'label': 'Stock DRS filesource', 'id': 'drsstock', 'doc': 'Make sure to define this generic drs file source if you have defined any other drs file sources, or stock drs download capability will be disabled.'}
    {'type': 'inveniordm', 'id': 'invenio_sandbox', 'doc': 'This is the Sandbox instance of Invenio. It is used for testing purposes only, content is NOT preserved. DOIs created in this instance are not real and will not resolve.', 'label': 'Invenio RDM Sandbox Repository (TESTING ONLY)', 'url': 'https://inveniordm.web.cern.ch/', 'token': "${user.user_vault.read_secret('preferences/invenio_sandbox/token')}", 'public_name': "${user.preferences['invenio_sandbox|public_name']}", 'writable': True}
    {'type': 'zenodo', 'id': 'zenodo', 'doc': 'Zenodo is a general-purpose open-access repository developed under the European OpenAIRE program and operated by CERN. It allows researchers to deposit data sets, research software, reports, and any other research-related digital artifacts. For each submission, a persistent digital object identifier (DOI) is minted, which makes the stored items easily citeable.', 'label': 'Zenodo', 'url': 'https://zenodo.org', 'token': "${user.user_vault.read_secret('preferences/zenodo/token')}", 'public_name': "${user.preferences['zenodo|public_name']}", 'writable': True}
    {'type': 'zenodo', 'id': 'zenodo_sandbox', 'doc': 'This is the Sandbox instance of Zenodo. It is used for testing purposes only, content is NOT preserved. DOIs created in this instance are not real and will not resolve.', 'label': 'Zenodo Sandbox (TESTING ONLY)', 'url': 'https://sandbox.zenodo.org', 'token': "${user.user_vault.read_secret('preferences/zenodo_sandbox/token')}", 'public_name': "${user.preferences['zenodo_sandbox|public_name']}", 'writable': True}
    {'type': 'onedata', 'id': 'onedata1', 'label': 'Onedata', 'doc': 'Your Onedata files - configure an access token via user preferences', 'access_token': "${user.preferences['onedata|access_token']}", 'onezone_domain': "${user.preferences['onedata|onezone_domain']}", 'disable_tls_certificate_validation': "${user.preferences['onedata|disable_tls_certificate_validation']}"}

   to

   ::

    {'type': 'dropbox', 'id': 'dropbox1', 'label': 'Dropbox files (configure access in user preferences)', 'doc': 'Your Dropbox files - configure an access token via the user preferences', 'access_token': "${user.preferences['dropbox|access_token']}"}
    {'type': 'webdav', 'id': 'owncloud1', 'label': 'OwnCloud', 'doc': 'External OwnCloud files (configure access in user preferences)', 'url': "${user.preferences['owncloud|url']}", 'root': "${user.preferences['owncloud|root']}", 'login': "${user.preferences['owncloud|username']}", 'password': "${user.preferences['owncloud|password']}", 'temp_path': '/your/temp/path', 'writable': False}
    {'type': 'posix', 'root': '/data/5/galaxy_import/galaxy_user_data/covid-19/data/sequences/', 'id': 'covid19-raw-sequences', 'label': 'COVID-19 FASTQ', 'doc': 'COVID-19 RAW sequences in FASTQ format'}
    {'type': 'posix', 'root': '/data/db/databases/pdb/pdb/', 'id': 'pdb-gzip', 'doc': 'Protein Data Bank (PDB)', 'label': 'PDB'}
    {'type': 'ftp', 'id': 'ebi-ftp', 'label': 'EBI FTP server', 'doc': 'European Bioinformatic Institute FTP server', 'host': 'ftp.ebi.ac.uk', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ftp', 'id': 'ncbi-ftp', 'label': 'NCBI FTP server', 'doc': 'NCBI FTP server', 'host': 'ftp.ncbi.nlm.nih.gov', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ftp', 'id': 'ensembl-ftp', 'label': 'ENSEMBL FTP server', 'doc': 'ENSEMBL FTP server', 'host': 'ftp.ensemblgenomes.org/vol1/pub/', 'user': 'anonymous', 'passwd': '', 'timeout': 10, 'port': 21}
    {'type': 'ssh', 'id': 'writeable-ssh-dir', 'requires_roles': 'writeable-galaxy-role', 'writable': True, 'label': 'Push your files to me', 'doc': 'This is an example of a writeable SSH dir', 'host': 'coolhost', 'user': 'user', 'passwd': 'passwd', 'timeout': 10, 'path': '/home/cooluser/', 'config_path': '', 'port': 2222}
    {'type': 's3fs', 'label': 'My MinIO storage', 'endpoint_url': 'https://minio.usegalaxy.eu', 'id': 'galaxy-minio-storage', 'doc': 'Galaxy MinIO S3 storage', 'anon': False, 'secret': 'UHAJ6asd6asdhasd', 'key': 'MCJU76agdt98GGFAROIP7'}
    {'type': 's3fs', 'label': 'Genome Ark', 'id': 'genomeark', 'doc': 'Access to Genome Ark open data on AWS.', 'bucket': 'genomeark', 'anon': True}
    {'type': 's3fs', 'label': '1000 Genomes', 'id': '1000genomes', 'doc': 'Access to the 1000 Genomes Project with human genetic variation, including SNPs, structural variants, and their haplotype context.', 'bucket': '1000genomes', 'anon': True}
    {'type': 's3fs', 'label': 'The Cancer Genome Atlas', 'id': 'tcga-2-open', 'doc': 'Access to the Cancer Genome Atlas (TCGA)', 'bucket': 'tcga-2-open', 'anon': True}
    {'type': 's3fs', 'label': 'COVID-19 Data Lake', 'id': 'covid19-lake', 'doc': 'A centralized repository of up-to-date and curated datasets on or related to the spread and characteristics of the novel corona virus (SARS-CoV-2) and its associated illness, COVID-19', 'bucket': 'covid19-lake', 'anon': True}
    {'type': 's3fs', 'label': 'Encyclopedia of DNA Elements (ENCODE)', 'id': 'encode-public', 'doc': 'The Encyclopedia of DNA Elements (ENCODE) Consortium is an international collaboration of research groups funded by the National Human Genome Research Institute (NHGRI)', 'bucket': 'encode-public', 'anon': True}
    {'type': 's3fs', 'label': 'Sentinel-3', 'id': 'meeo-s3-nrt', 'doc': 'European Commission’s Copernicus Earth Observation Programme. Sentinel-3 is a polar orbiting satellite that completes 14 orbits of the Earth a day.', 'bucket': 'meeo-s3/NRT/', 'anon': True}
    {'type': 's3fs', 'label': 'Sentinel-5P Level 2', 'id': 'meeo-s5p-nrti', 'doc': 'Observations from the Sentinel-5 Precursor satellite of the Copernicus Earth Observation Programme. It contains a polar orbiting satellite that completes 14 orbits of the Earth a day.', 'bucket': 'meeo-s5p/RPRO/', 'anon': True}
    {'type': 's3fs', 'label': 'Coupled Model Intercomparison Project 6', 'id': 'esgf-world', 'doc': 'The sixth phase of global coupled ocean-atmosphere general circulation model ensemble', 'bucket': 'esgf-world', 'anon': True}
    {'type': 's3fs', 'label': 'CMIP6 GCMs downscaled using WRF', 'id': 'wrf-cmip6-noversioning', 'doc': 'High-resolution historical and future climate simulations from 1980-2100', 'bucket': 'wrf-cmip6-noversioning', 'anon': True}
    {'type': 's3fs', 'label': 'NOAA Global Forecast System (GFS)', 'id': 'noaa-gfs-bdp-pds', 'doc': 'The Global Forecast System (GFS) is a weather forecast model produced by the National Centers for Environmental Prediction (NCEP).', 'bucket': 'noaa-gfs-bdp-pds', 'anon': True}
    {'type': 's3fs', 'label': 'NOAA Unified Forecast System Subseasonal to Seasonal Prototype 5', 'id': 'noaa-ufs-prototype5-pds', 'doc': 'The Unified Forecast System Subseasonal to Seasonal prototype 5 (UFS S2Sp5) dataset is reforecast data from the UFS atmosphere-ocean.', 'bucket': 'noaa-ufs-prototype5-pds', 'anon': True}
    {'type': 's3fs', 'label': 'Copernicus Digital Elevation Model (DEM)', 'id': 'copernicus-dem-30m', 'doc': 'The Copernicus DEM is a Digital Surface Model (DSM) which represents the surface of the Earth including buildings, infrastructure and vegetation.', 'bucket': 'copernicus-dem-30m', 'anon': True}
    {'type': 'http', 'label': 'Custom http filesource', 'id': 'httpcustom', 'url_regex': '^https?://myprotectedsite.org/', 'http_headers': {'Authorization': '#import base64\nBasic ${base64.b64encode(str.encode(user.preferences[\'myprotectedsite|username\'] + ":" + user.preferences[\'myprotectedsite|password\'])).decode()}'}, 'doc': 'Only define this if you want custom control over http downloads. You should also define a stock http source (below) or only downloads from this site will be allowed.'}
    {'type': 'http', 'label': 'Stock http filesource', 'id': 'httpstock', 'doc': 'Make sure to define this generic http file source if you have defined any other http file sources, or stock http download capability will be disabled.'}
    {'type': 'drs', 'label': 'Custom DRS filesource', 'id': 'drscustom', 'url_regex': '^drs://mydrssite.org/', 'http_headers': {'Authorization': '#import base64\nBasic ${base64.b64encode(str.encode(user.preferences[\'mydrssite|username\'] + ":" + user.preferences[\'mydrssite|password\'])).decode()}'}, 'doc': 'Define this if you want custom control over drs downloads. You should also define a stock drs source (below) or only downloads from this drs server will be allowed.'}
    {'type': 'drs', 'label': 'Stock DRS filesource', 'id': 'drsstock', 'doc': 'Make sure to define this generic drs file source if you have defined any other drs file sources, or stock drs download capability will be disabled.'}
    {'type': 'inveniordm', 'id': 'invenio_sandbox', 'doc': 'This is the Sandbox instance of Invenio. It is used for testing purposes only, content is NOT preserved. DOIs created in this instance are not real and will not resolve.', 'label': 'Invenio RDM Sandbox Repository (TESTING ONLY)', 'url': 'https://inveniordm.web.cern.ch/', 'token': "${user.user_vault.read_secret('preferences/invenio_sandbox/token')}", 'public_name': "${user.preferences['invenio_sandbox|public_name']}", 'writable': True}
    {'type': 'zenodo', 'id': 'zenodo', 'doc': 'Zenodo is a general-purpose open-access repository developed under the European OpenAIRE program and operated by CERN. It allows researchers to deposit data sets, research software, reports, and any other research-related digital artifacts. For each submission, a persistent digital object identifier (DOI) is minted, which makes the stored items easily citeable.', 'label': 'Zenodo', 'url': 'https://zenodo.org', 'token': "${user.user_vault.read_secret('preferences/zenodo/token')}", 'public_name': "${user.preferences['zenodo|public_name']}", 'writable': True}
    {'type': 'zenodo', 'id': 'zenodo_sandbox', 'doc': 'This is the Sandbox instance of Zenodo. It is used for testing purposes only, content is NOT preserved. DOIs created in this instance are not real and will not resolve.', 'label': 'Zenodo Sandbox (TESTING ONLY)', 'url': 'https://sandbox.zenodo.org', 'token': "${user.user_vault.read_secret('preferences/zenodo_sandbox/token')}", 'public_name': "${user.preferences['zenodo_sandbox|public_name']}", 'writable': True}
    {'type': 'dataverse', 'id': 'dataverse', 'doc': 'Dataverse is an open-source data repository platform designed for sharing, preserving, and managing research data, offering tools for data citation, exploration, and collaboration.', 'label': 'Dataverse', 'url': 'https://dataverse.org', 'token': "${user.user_vault.read_secret('preferences/dataverse/token')}", 'public_name': "${user.preferences['dataverse|public_name']}", 'writable': True}
    {'type': 'dataverse', 'id': 'dataverse_sandbox', 'doc': 'This is the sandbox instance of Dataverse. It is used for testing purposes only, content is NOT preserved. DOIs created in this instance are not real and will not resolve.', 'label': 'Dataverse Sandbox (use only for testing purposes)', 'url': 'https://demo.dataverse.org', 'token': "${user.user_vault.read_secret('preferences/dataverse_sandbox/token')}", 'public_name': "${user.preferences['dataverse_sandbox|public_name']}", 'writable': True}
    {'type': 'onedata', 'id': 'onedata1', 'label': 'Onedata', 'doc': 'Your Onedata files - configure an access token via user preferences', 'access_token': "${user.preferences['onedata|access_token']}", 'onezone_domain': "${user.preferences['onedata|onezone_domain']}", 'disable_tls_certificate_validation': "${user.preferences['onedata|disable_tls_certificate_validation']}"}
    {'type': 'elabftw', 'id': 'elabftw', 'label': 'eLabFTW', 'doc': 'Import/export files from an eLabFTW instance.', 'api_key': "${user.user_vault.read_secret('preferences/elabftw/api_key')}", 'writable': True, 'endpoint': "${user.preferences['elabftw|endpoint']}"}



config/galaxy.yml.sample:galaxy
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  mapping.galaxy.mapping.enable_tool_document_cache.desc has changed from

   ::

    Whether to enable the tool document cache. This cache stores
    expanded XML strings. Enabling the tool cache results in slightly faster startup
    times. The tool cache is backed by a SQLite database, which cannot
    be stored on certain network disks. The cache location is configurable
    with the ``tool_cache_data_dir`` tag in tool config files.

   to

   ::

    This option is deprecated, and the tool document cache will be removed
    in the next release.
    Whether to enable the tool document cache. This cache stores
    expanded XML strings. Enabling the tool cache results in slightly faster startup
    times. The tool cache is backed by a SQLite database, which cannot
    be stored on certain network disks. The cache location is configurable
    with the ``tool_cache_data_dir`` tag in tool config files.


-  mapping.galaxy.mapping.helpsite_url.required has changed from

   ::

    True

   to

   ::

    False




Removed
-------

The following configuration options have been completely removed

config/user_preferences_extra_conf.yml.sample
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  preferences.use_cached_job

config/galaxy.yml.sample:galaxy
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

-  mapping.galaxy.mapping.openai_api_key
-  mapping.galaxy.mapping.openai_model


Deprecation Notices
===========================================================

**Deprecation of the tool document cache in Galaxy release 25.1**
  The ``ToolDocumentCache`` (enabled via the config option: ``enable_tool_document_cache``) is
  deprecated and will be removed in the next Galaxy release (25.1). It is based on the unmaintained
  `sqlitedict <https://github.com/piskvorky/sqlitedict>`__ library and was not enabled on any of the
  UseGalaxy.* servers. Its only known usage was within AnVIL, which has since dropped support for it in
  `galaxyproject/galaxy-helm#438 <https://github.com/galaxyproject/galaxy-helm/pull/438>`__.

Developer Notes
===========================================================
Add content or drop section.

Release Notes
===========================================================

.. include:: 25.0.rst
   :start-after: announce_start

Release Team
===========================================================

Release manager:  `Ahmed Awan <https://github.com/ahmedhamidawan>`__, `John Davis <https://github.com/jdavcs>`__

Release testing:

* `Paulo Cilas Morais Lyra Junior <https://github.com/paulocilasjr>`__
* `Tyler Collins <https://github.com/tcollins2011>`__
* `Jayadev Joshi <https://github.com/jaidevjoshi83>`__
* `Jennifer Hillman-Jackson <https://github.com/jennaj>`__
* `Fabio Cumbo <https://github.com/cumbof>`__
* `Boris Depoortere <https://github.com/B0r1sD>`__
* `Michelle Savage <https://github.com/hujambo-dunia>`__
* `Daniela Schneider <https://github.com/Sch-Da>`__
* `Dannon Baker <https://github.com/dannon>`__
* `Mina Hojat Ansari <https://github.com/Minamehr>`__

A special thank you goes to everyone who helped test the new release after its deployment on usegalaxy.org.

----

.. include:: _thanks.rst
